src/ethereum_optimized/byzantium/state_db.py

```diff
@@ -117,749 +75,210 @@ def __eq__(self, other: object) -> bool:
+        state.db.destroy_storage(address)
+    for address, account in state.dirty_accounts.items():
+        state.db.set_account(address, account)
+    for address, storage in state.dirty_storage.items():
+        for key, value in storage.items():
+            state.db.set_storage(address, key, value)
+    state.destroyed_accounts = set()
+    state.dirty_accounts = {}
+    state.dirty_storage = {}
 
 
-def set_storage(
-    state: State, address: Address, key: Bytes, value: U256
-) -> None:
+def rollback_db_transaction(state: State) -> None:
     """
-    See `ethereum.byzantium.state`.
+    Rollback the current database transaction.
     """
-    if address not in state._dirty_accounts[-1]:
-        state._dirty_accounts[-1][address] = get_account_optional(
-            state, address
-        )
-    if address not in state._dirty_storage[-1]:
-        state._dirty_storage[-1][address] = {}
-    if value == 0:
-        state._dirty_storage[-1][address][key] = None
-    else:
-        state._dirty_storage[-1][address][key] = value
+    if len(state.tx_restore_points) != 0:
+        raise Exception("In a non-db transaction")
+    state.db.rollback_mutable()
+    state.dirty_accounts = {}
+    state.dirty_storage = {}
+    state.destroyed_accounts = set()
 
 
-def get_account_optional(state: State, address: Address) -> Optional[Account]:
+def begin_transaction(state: State) -> None:
     """
     See `ethereum.byzantium.state`.
     """
-    for cache in reversed(state._dirty_accounts):
-        if address in cache:
-            return cache[address]
-    internal_address = get_internal_key(address)
-    res = state._current_tx.get(b"\x01" + internal_address)
-    if res is None:
-        return None
-    else:
-        data = rlp.decode(res)
-        assert isinstance(data, list)
-        return Account(
-            Uint.from_be_bytes(data[0]), U256.from_be_bytes(data[1]), data[2]
-        )
+    if len(state.tx_restore_points) == 0:
+        flush(state)
+    state.tx_restore_points.append(len(state.journal))
 
 
-def get_account(state: State, address: Address) -> Account:
+def commit_transaction(state: State) -> None:
     """
     See `ethereum.byzantium.state`.
     """
-    res = get_account_optional(state, address)
-    if res is None:
-        return EMPTY_ACCOUNT
-    else:
-        return res
+    state.tx_restore_points.pop()
+    if len(state.tx_restore_points) == 0:
+        state.journal = []
+        flush(state)
 
 
-def set_account(
-    state: State, address: Address, account: Optional[Account]
-) -> None:
+def rollback_transaction(state: State) -> None:
     """
     See `ethereum.byzantium.state`.
     """
-    if account is None:
-        state._dirty_accounts[-1][address] = None
-    else:
-        state._dirty_accounts[-1][address] = account
+    restore_point = state.tx_restore_points.pop()
+    while len(state.journal) > restore_point:
+        item = state.journal.pop()
+        if len(item) == 3:
```

Comment:
Hm... I'd like to either see proper types for each of these branches `isinstance(item, SomeDescriptiveType)` or comments describing what each branch represents.